router/customdestinationmanager/customdestinationmanager.go

```diff
@@ -128,71 +140,73 @@ func (customManager *CustomManagerT) SendData(jsonData json.RawMessage, destID s
 		return 200, `200: outgoing disabled`
 	}
 
-	customManager.configSubscriberLock.RLock()
-	destLock, ok := customManager.destinationLockMap[destID]
-	customManager.configSubscriberLock.RUnlock()
+	customManager.mapLock.RLock()
+	clientLock, ok := customManager.clientLock[destID]
+	customManager.mapLock.RUnlock()
 	if !ok {
 		return 500, fmt.Sprintf("[CDM %s] Unexpected state: Lock missing for %s. Config might not have been updated. Please wait for a min before sending events.", customManager.destType, destID)
 	}
 
-	destLock.RLock()
-	customDestination, ok := customManager.destinationsMap[destID]
+	clientLock.RLock()
+	customDestination, ok := customManager.client[destID]
 
 	if !ok {
-		destLock.RUnlock()
-		destLock.Lock()
-		err := customManager.newClient(destID)
-		destLock.Unlock()
+		clientLock.RUnlock()
+		clientLock.Lock()
+		var err error
+		if _, ok := customManager.client[destID]; !ok {
```

Comment:
I was just going over the code to improve my understanding of the codebase. I was wondering if it would be better to just acquire the write lock while trying to access the map and release it upon finding the destination (or creating the client if required). I understand this might increase the time spent around this piece of code while the various coroutines are waiting for their turn to access the map, but assuming most of the time is spent in the network i/o this might not be creating any inefficiencies (and the code is much more readable)

Reply:
@raman2805 yes, you propose to use pessimistic locking, i.e. acquire a full lock from start, even if there is no need for it.
I chose optimistic locking instead, because I consider that the client is most likely to exist.
Provided that after a client gets successfully created it will remain there for all future calls, optimistic locking (and its associated performance advantages) appears to be a better choice.